File: architecture-skills/design-patterns/examples/mediator/mediator.java
// Mediator pattern in Java.
//
// The chat room centralizes routing so users never reference each other.
// The mediator interface lets colleagues be tested against a fake room.

import java.util.LinkedHashMap;
import java.util.Map;

public class mediator {

    interface ChatMediator {
        void broadcast(User sender, String text);
        void direct(User sender, String recipient, String text);
    }

    static final class ChatRoom implements ChatMediator {
        private final Map<String, User> users = new LinkedHashMap<>();

        void register(User user) {
            users.put(user.name(), user);
            user.join(this);
        }

        @Override
        public void broadcast(User sender, String text) {
            for (User user : users.values()) {
                if (user != sender) {
                    user.receive(sender.name(), text);
                }
            }
        }

        @Override
        public void direct(User sender, String recipient, String text) {
            User user = users.get(recipient);
            if (user == null) {
                sender.receive("room", "no such user: " + recipient);
            } else {
                user.receive(sender.name(), "(dm) " + text);
            }
        }
    }

    static final class User {
        private final String name;
        private ChatMediator room;

        User(String name) {
            this.name = name;
        }

        String name() {
            return name;
        }

        void join(ChatMediator room) {
            this.room = room;
        }

        void say(String text) {
            room.broadcast(this, text);
        }

        void dm(String recipient, String text) {
            room.direct(this, recipient, text);
        }

        void receive(String sender, String text) {
            System.out.printf("[%s] %s: %s%n", name, sender, text);
        }
    }

    public static void main(String[] args) {
        ChatRoom room = new ChatRoom();
        User ada = new User("Ada");
        User grace = new User("Grace");
        User alan = new User("Alan");
        room.register(ada);
        room.register(grace);
        room.register(alan);

        ada.say("Hello, everyone!");
        grace.dm("Ada", "Hi Ada, got a minute?");
        alan.dm("Linus", "ping");
    }
}

File: architecture-skills/design-patterns/examples/mediator/mediator.php
<?php
// Mediator pattern in PHP.
//
// The chat room centralizes routing so users never reference each other.
// Frameworks often provide this shape as an event dispatcher
// (e.g. Symfony EventDispatcher) acting as the mediator.

declare(strict_types=1);

final class ChatRoom
{
    /** @var array<string, User> */
    private array $users = [];

    public function register(User $user): void
    {
        $this->users[$user->name] = $user;
        $user->join($this);
    }

    public function broadcast(User $sender, string $text): void
    {
        foreach ($this->users as $user) {
            if ($user !== $sender) {
                $user->receive($sender->name, $text);
            }
        }
    }

    public function direct(User $sender, string $recipient, string $text): void
    {
        $user = $this->users[$recipient] ?? null;
        if ($user === null) {
            $sender->receive('room', "no such user: $recipient");
        } else {
            $user->receive($sender->name, "(dm) $text");
        }
    }
}

final class User
{
    private ?ChatRoom $room = null;

    public function __construct(public readonly string $name)
    {
    }

    public function join(ChatRoom $room): void
    {
        $this->room = $room;
    }

    public function say(string $text): void
    {
        $this->room->broadcast($this, $text);
    }

    public function dm(string $recipient, string $text): void
    {
        $this->room->direct($this, $recipient, $text);
    }

    public function receive(string $sender, string $text): void
    {
        echo "[{$this->name}] $sender: $text\n";
    }
}

$room = new ChatRoom();
$ada = new User('Ada');
$grace = new User('Grace');
$alan = new User('Alan');
foreach ([$ada, $grace, $alan] as $user) {
    $room->register($user);
}

$ada->say('Hello, everyone!');
$grace->dm('Ada', 'Hi Ada, got a minute?');
$alan->dm('Linus', 'ping');
